circuits/cpp/src/aztec3/circuits/abis/function_leaf_preimage.hpp

```diff
@@ -32,14 +34,15 @@ template <typename NCT> struct FunctionLeafPreimage {
     using uint32 = typename NCT::uint32;
 
     uint32 function_selector = 0;
+    boolean is_internal = false;
```

Comment:
Possibly, 3 are overlapping. But the function data is accesible inside noir as well, so maybe the leaf should include the function data: `leaf = {function_data, vk_hash, acir_hash}`. Seems more intuitive to have the `is_constructor` stored in the leaf as well. Don't recall how we right now are ensuring the constructors are not rerun @iAmMichaelConnor.

Reply:
I like the sounds of that

Reply:
I think perhaps the difference between `function_data` and `function_leaf_preimage` (at least, when things were initially designed... having quickly looked back at the diagram, which is basically an extension of my memory at this point)... is...

`function_data` is information about the function you wish to call, which becomes part of a tx_request. It's separate from any deployment information. Perhaps it should be renamed to `target_function` or `entrypoint_function` or something. Given this information, a node which has already synced can lookup the rest of the function's info (vk_hash, acir_hash, etc).

`function_leaf_preimage` is data required to prove existence of the function in the contracts tree. Some of that data comes from `function_data`, and some of that data is grabbed by the rpc server.

Some of this is faint memories, so I could be wrong, but I believe their separation was intentional.

Re storing `is_constructor` in the leaf. In the initial design, the constructor isn't included as a leaf in the tree (because it never needs to be called after deployment); so the constructor's information is only included in the `constructor_hash` which is included in the contract's address. I'm not sure if that's still the case in the code. I intend to write lots of specs soon, and we'll review all this.

> Don't recall how we right now are ensuring the constructors are not rerun
We emit a nullifier = h(contract_address). You can only call a constructor at the time of deployment, and at deployment the nullifier gets emitted, so there's no way to call a constructor twice.

Reply:
sounds good, it can stay as is!